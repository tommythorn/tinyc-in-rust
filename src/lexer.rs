#![warn(clippy::all, clippy::pedantic)]

/* Copyright (C) 2023 by Tommy Thorn, All Rights Reserved. */

/* Lexer. */

/// Source code is first broken into atomic `Token`s which are the
/// only thing the `Parser` seens.
#[derive(Clone, Debug)]
pub enum Token {
    DoSym,
    ElseSym,
    IfSym,
    WhileSym,
    Lbra,
    Rbra,
    Lpar,
    Rpar,
    Plus,
    Minus,
    Less,
    Semi,
    Equal,
    Int(isize),
    Id(String),
    Eoi,
}

/// Source code position
#[derive(Clone, Copy)]
struct Pos {
    line: usize,
    col: usize,
}

/// The `Lexer` is initialized with the source code string and
/// tokenizes it, keeping the current `Token` in `sym`.  We get a new
/// token by calling `next_sym()`.
///
/// NB: A much nicer lexer would track where in the source we are so
/// syntax errors could be more user friendly.
pub struct Lexer<'a> {
    /// The peekable iterator that gives us chars from the source
    itr: std::iter::Peekable<std::str::Chars<'a>>,

    /// The _current_ position (line, col) in the source
    pos: Pos,

    /// The position of the token in `sym`
    sym_pos: Pos,

    /// The current token (the "lookahead")
    pub sym: Token,
}

impl<'a> Lexer<'a> {
    #[must_use]
    pub fn new(src: &'a str) -> Lexer<'a> {
        let mut lex = Self {
            itr: src.chars().peekable(),
            pos: Pos { line: 1, col: 1 },
            sym_pos: Pos { line: 0, col: 0 },
            sym: Token::Eoi,
        };

        lex.next_sym();

        lex
    }

    /// Report a error message in the context of the current lexer position and terminate (panic)
    /// # Panics
    /// Sure does
    pub fn syntax_error(&mut self, msg: &str) {
        eprintln!("input:{}:{}:{msg}", self.sym_pos.line, self.sym_pos.col);
        // Proper error handling is out of scope for now.
        std::process::exit(1);
    }

    fn next_ch(&mut self) {
        self.itr.next();
        self.pos.col += 1;
        if self.ch() == '\n' {
            self.pos.line += 1;
            self.pos.col = 1;
        }
    }

    fn ch(&mut self) -> char {
        *self.itr.peek().unwrap_or(&'\0')
    }

    /// Parses the next `Token` and populates `self.sym` with it.
    /// `Token::Eoi` is represents the end of the source code.
    pub fn next_sym(&mut self) {
        while self.ch() == ' ' || self.ch() == '\n' {
            self.next_ch();
        }

        self.sym_pos = self.pos;

        match self.ch() {
            '\0' => self.sym = Token::Eoi,
            '{' => self.sym = Token::Lbra,
            '}' => self.sym = Token::Rbra,
            '(' => self.sym = Token::Lpar,
            ')' => self.sym = Token::Rpar,
            '+' => self.sym = Token::Plus,
            '-' => self.sym = Token::Minus,
            '<' => self.sym = Token::Less,
            ';' => self.sym = Token::Semi,
            '=' => self.sym = Token::Equal,

            '0'..='9' => {
                let mut int_val = 0;
                while '0' <= self.ch() && self.ch() <= '9' {
                    int_val = int_val * 10 + self.ch() as isize - '0' as isize;
                    self.next_ch();
                }

                self.sym = Token::Int(int_val);

                return;
            }

            'a'..='z' => {
                let mut id_name: String = String::new();
                while 'a' <= self.ch() && self.ch() <= 'z' || self.ch() == '_' {
                    id_name.push(self.ch());
                    self.next_ch();
                }

                self.sym = match id_name.as_str() {
                    "do" => Token::DoSym,
                    "else" => Token::ElseSym,
                    "if" => Token::IfSym,
                    "while" => Token::WhileSym,
                    _ => Token::Id(id_name),
                };

                return;
            }

            _ => self.syntax_error("Illegal token"),
        }

        self.next_ch();
    }
}
