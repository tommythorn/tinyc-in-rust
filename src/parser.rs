//! The parser for Tiny-C in classic recursive-descent style
//!
//! Normally in C we would use global variables for the current token,
//! source code position, etc, but in Rust we wrap everything in a
//! struct and implement the parser functions as (recursive) member
//! functions.

#![warn(clippy::all, clippy::pedantic)]

use crate::lexer::{Lexer, Token};

/// To create recursive types in Rust, we heap allocate the recursive
/// subparts, via the `Box` type.  To keep the `Node` type more
/// readable we use an alias for the boxed node.
pub type BNode = Box<Node>;

/// The `Node` is the abstract syntax tree.  This would normally be
/// segregated into the syntatic categories like expression,
/// statement, etc., but for this little example we just bundle
/// everything, forgoing a bit of type safety for brevity.
#[derive(Debug)]
pub enum Node {
    /// Contains the named variable
    Var(String),

    /// Contains integer constants
    Cst(isize),

    /// An addition expression
    Add(BNode, BNode),

    /// A subtraction expression
    Sub(BNode, BNode),

    /// A less-than boolean expression
    Lt(BNode, BNode),

    /// The assignment statement.  Note, the first argument must be `Var(_)`.
    Set(BNode, BNode),

    /// An `if` statement with no `else` part.
    If1(BNode, BNode),

    /// An `if` statement with an `else` part.
    If2(BNode, BNode, BNode),

    /// A `while` statement with test and body
    While(BNode, BNode),

    /// A `do-while` statement with body and test
    Do(BNode, BNode),

    /// The null statement, for compiler convenience
    Empty,

    /// The sequence node, which ties together two (or more) statements
    Seq(BNode, BNode),

    /// The expression statement
    Expr(Box<Node>),

    /// The top-level program (there should be exactly one of these)
    Prog(Box<Node>),
}

/// The main entry point to the parser
///
/// ```
/// use tinyc_in_rust::parser::{Node,parse};
/// let ast: Node = parse("q = 42;");
/// ```
#[must_use]
pub fn parse(src: &str) -> Node {
    Parser::new(src).program()
}

/// The `Parser` parses a source string into a `Node` tree
/// representation
struct Parser<'a> {
    lex: Lexer<'a>,
}

impl<'a> Parser<'a> {
    /// Prepare for parsing, given the provided source code
    fn new(src: &'a str) -> Self {
        Self {
            lex: Lexer::new(src),
        }
    }

    /// Returns the current token (also know as the "lookahead")
    fn sym(&self) -> Token {
        self.lex.sym.clone()
    }

    /// Fetches the next token from the lexer
    fn next_sym(&mut self) {
        self.lex.next_sym();
    }

    /// Parser for the `<term>` syntax
    /// `<term> ::= <id> | <int> | <paren_expr>`
    fn term(&mut self) -> Node {
        match self.sym() {
            Token::Id(name) => {
                self.next_sym();
                Node::Var(name)
            }
            Token::Int(val) => {
                self.next_sym();
                Node::Cst(val)
            }
            _ => self.paren_expr(),
        }
    }

    /* <sum> ::= <term> | <sum> "+" <term> | <sum> "-" <term> */
    fn sum(&mut self) -> Node {
        let mut t = self.term();
        loop {
            if matches!(self.sym(), Token::Plus) {
                self.next_sym();
                t = Node::Add(Box::new(t), Box::new(self.term()));
            } else if matches!(self.sym(), Token::Minus) {
                self.next_sym();
                t = Node::Sub(Box::new(t), Box::new(self.term()));
            } else {
                return t;
            }
        }
    }

    /* <test> ::= <sum> | <sum> "<" <sum> */
    fn cond(&mut self) -> Node {
        let l = self.sum();
        if matches!(self.sym(), Token::Less) {
            self.next_sym();
            Node::Lt(Box::new(l), Box::new(self.sum()))
        } else {
            l
        }
    }

    /* <expr> ::= <test> | <id> "=" <expr> */
    fn expr(&mut self) -> Node {
        if !matches!(self.sym(), Token::Id(_)) {
            return self.cond();
        }
        let t = self.cond(); // == Node::Var(..)
        if matches!(self.sym(), Token::Equal) {
            self.next_sym();
            Node::Set(Box::new(t), Box::new(self.expr()))
        } else {
            t
        }
    }

    fn paren_expr(&mut self) -> Node {
        if !matches!(self.sym(), Token::Lpar) {
            self.lex.syntax_error("`(' expected");
        }
        self.next_sym();
        let x = self.expr();
        if !matches!(self.sym(), Token::Rpar) {
            self.lex.syntax_error("`)' expected");
        }
        self.next_sym();

        x
    }

    fn statement(&mut self) -> Node {
        if matches!(self.sym(), Token::IfSym) {
            /* "if" <paren_expr> <statement> */
            self.next_sym();
            let cond = self.cond();
            let then = self.statement();
            if matches!(self.sym(), Token::ElseSym) {
                /* ... "else" <statement> */
                self.next_sym();
                Node::If2(Box::new(cond), Box::new(then), Box::new(self.statement()))
            } else {
                Node::If1(Box::new(cond), Box::new(then))
            }
        } else if matches!(self.sym(), Token::WhileSym) {
            /* "while" <paren_expr> <statement> */
            self.next_sym();
            let cond = self.paren_expr();
            Node::While(Box::new(cond), Box::new(self.statement()))
        } else if matches!(self.sym(), Token::DoSym) {
            /* "do" <statement> "while" <paren_expr> ";" */
            self.next_sym();
            let body = self.statement();
            if !matches!(self.sym(), Token::WhileSym) {
                self.lex.syntax_error("expected `while'");
            }
            self.next_sym();
            let cond = self.paren_expr();
            if !matches!(self.sym(), Token::Semi) {
                self.lex.syntax_error("expected `;'");
            }
            self.next_sym();
            Node::Do(Box::new(body), Box::new(cond))
        } else if matches!(self.sym(), Token::Semi) {
            /* ";" */
            self.next_sym();
            Node::Empty
        } else if matches!(self.sym(), Token::Lbra) {
            /* "{" { <statement> } "}" */
            self.next_sym();
            let mut x = self.statement();
            while !matches!(self.sym(), Token::Rbra) {
                x = Node::Seq(Box::new(x), Box::new(self.statement()));
            }
            self.next_sym();
            x
        } else {
            /* <expr> ";" */
            let x = self.expr();
            if !matches!(self.sym(), Token::Semi) {
                self.lex.syntax_error("expected `;'");
            }
            self.next_sym();
            Node::Expr(Box::new(x))
        }
    }

    fn program(&mut self) -> Node {
        /* <program> ::= <statement> */
        let stmt = self.statement();
        if !matches!(self.sym(), Token::Eoi) {
            self.lex.syntax_error("program ended here");
        }
        Node::Prog(Box::new(stmt))
    }
}

// *** Parser Testing ***

#[cfg(test)]
use insta::assert_snapshot;

#[test]
fn test_term() {
    let mut parse = Parser::new("2 alpha");
    let n = parse.term();
    assert!(matches!(n, Node::Cst(2)));
    let n = parse.term();
    assert!(match n {
        Node::Var(v) => v == "alpha",
        _ => false,
    });
}

#[test]
fn test_sum() {
    assert_snapshot!(format!("{:?}", Parser::new("2+3-4").sum()));
    assert_snapshot!(format!("{:?}", Parser::new("a-b-c").sum()));
}

#[test]
fn test_cond() {
    assert_snapshot!(format!("{:?}", Parser::new("2 < 4").cond()));
    assert_snapshot!(format!("{:?}", Parser::new("a").cond()));
}

#[test]
fn test_expr() {
    assert_snapshot!(format!("{:?}", Parser::new("2 < 4").expr()));
    assert_snapshot!(format!("{:?}", Parser::new("a = 42 - 666").expr()));
}

#[test]
fn test_paren_expr() {
    assert_snapshot!(format!("{:?}", Parser::new("(2-(3-4))").paren_expr()));
    assert_snapshot!(format!("{:?}", Parser::new(" (x < 7) y;").paren_expr()));
}

#[test]
fn test_statement() {
    assert_snapshot!(format!("{:?}", Parser::new(";").statement()));
    assert_snapshot!(format!("{:?}", Parser::new("a;").statement()));
    assert_snapshot!(format!(
        "{:?}",
        Parser::new("if (2 < 3) b = 42;").statement()
    ));
}

#[test]
fn test_statement2() {
    assert_snapshot!(format!(
        "{:?}",
        Parser::new("if (2) b = 42; else b = 666;").statement()
    ));
}

#[test]
fn test_statement3() {
    assert_snapshot!(format!(
        "{:?}",
        Parser::new("{ b = 666; c = 3; d = b; }").statement()
    ));
}

#[test]
fn test_statement4() {
    assert_snapshot!(format!("{:?}", Parser::new("while (x < 7) y;").statement()));
    assert_snapshot!(format!(
        "{:?}",
        Parser::new("while (x < 7) { b = b - 1; c = c + b; }").statement()
    ));
}

#[test]
fn test_program() {
    assert_snapshot!(format!("{:?}", Parser::new("a = 42;").program()));
}
