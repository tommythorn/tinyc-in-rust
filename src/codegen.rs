//! Compile from parsed source code to code as a list of instructions.

#![warn(clippy::all, clippy::pedantic)]

use crate::parser::Node;

/// `Insn` models the instructions of our virtual machine.
///
/// Marc Feeley uses fixed size for each instruction, with Push,
/// Fetch, Load, Jmp, Jz, and Jnz taking two slots.  We model the cast
/// of the slot values with the `Integer(_)` and `Address(_)` types
/// which of course wouldn't exist in a Real Implementation.  An
/// alternative (more type safe and more conventional) approach would
/// be to use `Fetch(usize)` etc. directly.
///
/// The targets of `Jmp`, `Jnz`, and `Jz` are absolute addresses.
/// Conventionally they would be relative addresses.
#[derive(Debug)]
pub enum Insn {
    Fetch,
    Store,
    Push,
    Pop,
    Add,
    Sub,
    Lt,
    Jz,
    Jnz,
    Jmp,
    Halt,
    Integer(isize),
    Address(usize),
}

/// Take the top-level program Node and compile it to instructions.
#[must_use]
pub fn compile(ast: Node) -> Vec<Insn> {
    let mut cg = Codegen::default();
    cg.compile(ast);
    cg.code
}

/// The Generator traverses the parsed source code and generates
/// `code` in the process.
#[derive(Default)]
struct Codegen {
    code: Vec<Insn>,
}

impl Codegen {
    #[allow(clippy::unused_self)]
    fn global(&self, v: &str) -> usize {
        match v {
            v if v.len() == 1 => v.chars().next().unwrap() as usize - 97,
            _ => panic!("{v} isn't a variable we can compile right now"),
        }
    }

    fn here(&self) -> usize {
        self.code.len()
    }

    fn hole(&mut self) -> usize {
        let p = self.here();
        self.code.push(Insn::Address(0));
        p
    }

    fn fix(&mut self, hole: usize, target: usize) {
        self.code[hole] = Insn::Address(target);
    }

    fn compile(&mut self, n: Node) {
        match n {
            Node::Add(a, b) => {
                self.compile(*a);
                self.compile(*b);
                self.code.push(Insn::Add);
            }
            Node::Sub(a, b) => {
                self.compile(*a);
                self.compile(*b);
                self.code.push(Insn::Sub);
            }
            Node::If1(test, then) => {
                self.compile(*test);
                self.code.push(Insn::Jz);
                let jz = self.hole();

                self.compile(*then);
                self.fix(jz, self.here());
            }
            Node::If2(test, then, else_) => {
                self.compile(*test);
                self.code.push(Insn::Jz);
                let jz = self.hole();

                self.compile(*then);
                self.code.push(Insn::Jmp);
                let jmp = self.hole();

                self.fix(jz, self.here());
                self.compile(*else_);

                self.fix(jmp, self.here());
            }
            Node::While(test, body) => {
                let l_restart = self.here();

                self.compile(*test);

                self.code.push(Insn::Jz);
                let jz = self.hole();

                self.compile(*body);
                self.code.push(Insn::Jmp);
                let jmp = self.hole();

                self.fix(jmp, l_restart);
                self.fix(jz, self.here());
            }
            Node::Do(body, test) => {
                let l_restart = self.here();

                self.compile(*body);
                self.compile(*test);

                self.code.push(Insn::Jnz);
                let jnz = self.hole();
                self.fix(jnz, l_restart);
            }
            Node::Prog(body) => {
                self.compile(*body);
                self.code.push(Insn::Halt);
            }
            Node::Expr(body) => {
                self.compile(*body);
                self.code.push(Insn::Pop);
            }
            Node::Set(var, expr) => {
                self.compile(*expr);
                self.code.push(Insn::Store);
                let Node::Var(v) = *var else {
                    panic!("We expected a Var, not {:?}", *var);
                };
                self.code.push(Insn::Address(self.global(&v)));
            }
            Node::Cst(val) => {
                self.code.push(Insn::Push);
                self.code.push(Insn::Integer(val));
            }
            Node::Var(v) => {
                self.code.push(Insn::Fetch);
                self.code.push(Insn::Address(self.global(&v)));
            }
            Node::Lt(a, b) => {
                self.compile(*a);
                self.compile(*b);
                self.code.push(Insn::Lt);
            }
            Node::Seq(a, b) => {
                self.compile(*a);
                self.compile(*b);
            }
            Node::Empty => {}
        }
    }
}
