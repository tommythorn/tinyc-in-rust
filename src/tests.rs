#![warn(clippy::all, clippy::pedantic)]
use crate::codegen::compile;
use crate::lexer::{Lexer, Token};
use crate::parser::parse;
use insta::assert_snapshot;

// *** Lexer Testing ***

#[test]
fn test_lexer() {
    let mut lex = Lexer::new("2 3 alpha beta ={}");
    assert!(matches!(lex.sym, Token::Int(2)));
    lex.next_sym();
    assert!(matches!(lex.sym, Token::Int(3)));
    lex.next_sym();
    assert!(match &lex.sym {
        Token::Id(v) => v == "alpha",
        _ => false,
    });
    lex.next_sym();
    assert!(match &lex.sym {
        Token::Id(v) => v == "beta",
        _ => false,
    });
    lex.next_sym();
    assert!(matches!(lex.sym, Token::Equal));
    lex.next_sym();
    assert!(matches!(lex.sym, Token::Lbra));
    lex.next_sym();
    assert!(matches!(lex.sym, Token::Rbra));
    lex.next_sym();
    assert!(matches!(lex.sym, Token::Eoi));
    lex.next_sym();
    assert!(matches!(lex.sym, Token::Eoi));
}

// *** Compiler Testing ***

fn show_code(src: &str) -> String {
    format!("{:?}", compile(parse(src)))
}

const EXAMPLES: [&str; 7] = [
    "a=b=c=2<3;",
    "{ i=1; while (i<100) i=i+i; }",
    "{ i=125; j=100; while (i-j) if (i<j) j=j-i; else i=i-j; }",
    "{ i=1; do i=i+10; while (i<50); }",
    "{ i=1; while ((i=i+10)<50) ; }",
    "{ i=7; if (i<5) x=1; if (i<10) y=2; }",
    "{ m=n=1;k=10; while (0 < k) { t = m; m = n; n = t + n; k = k - 1; }}",
];

#[test]
fn test_cg_assignment() {
    assert_snapshot!(show_code("a = 42;"));
}

#[test]
fn test_cg_examples() {
    for ex in EXAMPLES {
        assert_snapshot!(show_code(ex));
    }
}

// *** Execution Testing ***

#[test]
fn test_run_examples() {
    for ex in EXAMPLES {
        println!("Try {ex}:");
        crate::vm::VM::new().run(compile(parse(ex)));
    }
}
