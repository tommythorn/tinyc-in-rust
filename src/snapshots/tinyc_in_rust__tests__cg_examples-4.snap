---
source: src/tests.rs
expression: show_code(ex)
---
[Push, Integer(1), Store, Address(8), Pop, Fetch, Address(8), Push, Integer(10), Add, Store, Address(8), Pop, Fetch, Address(8), Push, Integer(50), Lt, Jnz, Address(5), Halt]
